### PR TITLE
Binder Docs + Performance improvements

### PR DESCRIPTION
**Pull request by varon**
### Purpose of this PR

Did this while reviewing #795 
This adds some documentation and simplifies the file writing parts.

### Testing status

Seems to work :)
The struct scope brackets and the removal of 10000 async calls in essentially synchronous methods helped get the binder performance to under 9 seconds on my PC, which is pretty old now.

### CODE REVIEW

File: src/Generator.Bind/Generators/Bases/GeneratorSettingsBase.cs [hunk 1]
```diff
@@ -40,6 +40,9 @@ internal abstract class GeneratorSettingsBase : IGeneratorSettings
         /// <inheritdoc/>
         public abstract string APITypemap { get; }
 
+        /// <inheritdoc/>
+        public abstract string LicenseText { get; }
```
Comment (Nihlus) [inline src/Generator.Bind/Generators/Bases/GeneratorSettingsBase.cs]: The settings class should not contain actual data - if a configurable license is required, it should be a path to a file instead. 

Beyond that, why is this needed? Was the license not inserted properly?

File: src/Generator.Bind/Translation/Mappers/TypeMapper.cs [hunk 2]
```diff
@@ -6,6 +6,7 @@ namespace Bind.Translation.Mappers
 {
     /// <summary>
     /// Maps OpenGL types to C# language types.
+    ///
```
Comment (Nihlus) [inline src/Generator.Bind/Translation/Mappers/TypeMapper.cs]: Spurious newline.

File: src/Generator.Bind/Writers/BindingWriter.cs [hunk 3]
```diff
@@ -142,7 +142,7 @@ private Task WriteWrappersAsync()
             var extensionNames = _profile.GetExtensions();
 
             // Create the interop setup necessary
-            var createConstructorFileTask = CreateWrapperClassConstructorFileAsync(wrappersOutputDirectory);
+            var createConstructorFileTask = Task.Run(() => CreateWrapperClassConstructorFileAsync(wrappersOutputDirectory));
```
Comment (Nihlus) [inline src/Generator.Bind/Writers/BindingWriter.cs]: Why is Task.Run used here?

File: src/Generator.Bind/Writers/BindingWriter.cs [hunk 4]
```diff
@@ -160,12 +160,12 @@ private Task WriteExtensionWrapperFiles([NotNull] string extensionName, [NotNull
...
 
             var allCategories = _profile.GetCategories();
-            var categoryTasks = allCategories.Select(category => WriteCategoryWrapperFile(extensionOutputDirectory, extensionName, category));
+            var categoryTasks = allCategories.Select(category => Task.Run(() => WriteCategoryWrapperFile(extensionOutputDirectory, extensionName, category)));
 
             return Task.WhenAll(categoryTasks);
         }
 
-        private async Task WriteCategoryWrapperFile
+        private void WriteCategoryWrapperFile
```
Comment (Nihlus) [inline src/Generator.Bind/Writers/BindingWriter.cs]: I'd like to see before and after numbers on this change. The intent is to allow asynchronous writing of multiple files at once, and this change makes this method strictly synchronous. 

With the proposed changes, I would expect performance to drop for multi-target runs.

File: src/Generator.Bind/XML/Signatures/Functions/FunctionSignature.cs [hunk 5]
```diff
@@ -69,7 +69,7 @@ public class FunctionSignature : INamedExtensionScopedEntity
         /// <summary>
         /// Gets a value indicating whether the function has generic type parameters.
         /// </summary>
-        public bool HasGenericTypeParameters => GenericTypeParameters.Any();
```
Comment (Nihlus) [inline src/Generator.Bind/XML/Signatures/Functions/FunctionSignature.cs]: Any particular reason for this? LINQ is usually cheap and more expressive. If it's for perf, I'd like to see before and after numbers.

File: src/Generator.Bind/XML/Signatures/SignatureReader.cs [hunk 6]
```diff
@@ -13,7 +13,9 @@
 namespace Bind.XML.Signatures
 {
     /// <summary>
-    /// Reads signature definitions from an XML document.
+    /// Reads API profile definitions from an XML document.
+    ///
```
Comment (Nihlus) [inline src/Generator.Bind/XML/Signatures/SignatureReader.cs]: The reader reads signature definitions, not API profile definitions. Different concepts. If the class doesn't do that, the name of the class should change.

Comment (Nihlus) [inline src/Generator.Bind/XML/Signatures/SignatureReader.cs]: The reader reads signature definitions, not API profile definitions. Different concepts. If the class does not do that, its name should change.

Comment (varon) [reply src/Generator.Bind/XML/Signatures/SignatureReader.cs]: Will revert. This is a 4am special.

File: src/Generator.Bind/Generators/Bases/GeneratorSettingsBase.cs (hunk 1 above)
Comment (varon) [reply src/Generator.Bind/Generators/Bases/GeneratorSettingsBase.cs]: The license text was being read from a file every single time we wrote a file. This is a huge amount of disk IO that we've now avoided by reading it once.

File: src/Generator.Bind/Writers/BindingWriter.cs (hunk 4 above)
Comment (varon) [reply src/Generator.Bind/Writers/BindingWriter.cs]: You can't parallelize writing to a single file; it's a fundamentally synchronous operation because the order of the text you're writing matters. What we can do is run multiple of these synchronous file IO operations as parallel tasks, via Task.Run.

By eliminating a bajillion task invocations via the await keyword, we're able to get faster performance within each of these tasks, and better tuning of our parallelism.

Some benchmarks:

Before:
```
Bind.exe
OpenGL binding generator 1.0.0.0 for OpenTK.
For comments, bugs and suggestions visit http://github.com/opentk/opentk

Writing bindings to src\OpenTK\Graphics
Writing bindings to src\OpenTK\Graphics
Writing bindings to src\OpenTK\Graphics
Writing bindings to src\OpenTK\Graphics
Writing bindings to src\OpenTK\Graphics
Writing bindings to src\OpenTK\Graphics
Writing bindings to src\OpenTK\Graphics

Bindings generated in 12.3194237 seconds.

Process finished with exit code 0.
```

After:

```
>Bind.exe
OpenGL binding generator 1.0.0.0 for OpenTK.
For comments, bugs and suggestions visit http://github.com/opentk/opentk

Writing bindings to src\OpenTK\Graphics
Writing bindings to src\OpenTK\Graphics
Writing bindings to src\OpenTK\Graphics
Writing bindings to src\OpenTK\Graphics
Writing bindings to src\OpenTK\Graphics
Writing bindings to src\OpenTK\Graphics
Writing bindings to src\OpenTK\Graphics

Bindings generated in 8.288314 seconds.

Process finished with exit code 0.
```

File: src/Generator.Bind/Writers/BindingWriter.cs (hunk 3 above)
Comment (varon) [reply src/Generator.Bind/Writers/BindingWriter.cs]: The method name should no longer be Async. Will fix.

File: src/Generator.Bind/XML/Signatures/Functions/FunctionSignature.cs (hunk 5 above)
Comment (varon) [reply src/Generator.Bind/XML/Signatures/Functions/FunctionSignature.cs]: While this may not be performance critical code, creating an iterator and then enumerating a sequence when you can do a field comparison is just crazy.

Remember, when expanded, `Any()` is doing ALL of this underneath the hood.

```csharp
if (GenericTypeParameters == null)
{
    throw new ArgumentNullException(nameof(GenericTypeParameters));
}

var e = GenericTypeParameters.GetEnumerator();
try
{
    return e.MoveNext();
}
finally
{
    e.Dispose();
}
```
vs the vastly simpler int comparison:
```csharp
GenericTypeParameters.Count == 0
````

File: src/Generator.Bind/XML/Signatures/SignatureReader.cs (hunk 6 above)
Comment (varon) [reply src/Generator.Bind/XML/Signatures/SignatureReader.cs]: Will fix.

File: src/Generator.Bind/Writers/BindingWriter.cs (hunk 4 above)
Comment (VPeruS) [reply src/Generator.Bind/Writers/BindingWriter.cs]: Why not Task WriteText(string[])?

Comment (varon) [reply src/Generator.Bind/Writers/BindingWriter.cs]: I actually tried using pre-allocated buffers and then writing the files at once. This surprisingly was slower due to the fact that each line had to be written twice now. It turns out that .Net is doing a really good job of buffering the file IO to make it as fast as possible.

File: src/Generator.Bind/Generators/Bases/GeneratorSettingsBase.cs (hunk 1 above)
Comment (Nihlus) [reply src/Generator.Bind/Generators/Bases/GeneratorSettingsBase.cs]: Ah, that's a good reason. However, it still doesn't belong in a settings class. It should be sourced somewhere else.

File: src/Generator.Bind/Writers/BindingWriter.cs (hunk 4 above)
Comment (Nihlus) [reply src/Generator.Bind/Writers/BindingWriter.cs]: Are these tests with or without the license text change? The idea isn't that a single file should be parallelized - it's obviously not, what with `await` all over the place - but rather, writing to different files in potentially different targets in parallel.

File: src/Generator.Bind/XML/Signatures/Functions/FunctionSignature.cs (hunk 5 above)
Comment (Nihlus) [reply src/Generator.Bind/XML/Signatures/Functions/FunctionSignature.cs]: I see the point, but I personally prefer the more semantically direct `Any()` call. The speed difference ought to be negligible.

File: src/Generator.Bind/Program.cs [hunk 7]
```diff
@@ -62,6 +63,8 @@ public static class Program
         /// <returns>An integer, indicating success or failure. On a failure, a nonzero value is returned.</returns>
         private static async Task<int> Main(string[] args)
         {
+            // force the GC off for a bit more speed.
+            GCSettings.LatencyMode = GCLatencyMode.Batch;
```
Comment (Nihlus) [inline src/Generator.Bind/Program.cs]: How much of a difference does this make? I'd prefer to avoid arcane code if the benefits are minimal :P

File: src/Generator.Bind/Writers/BindingWriter.cs [hunk 8]
```diff
@@ -175,15 +175,13 @@ private Task WriteExtensionWrapperFiles([NotNull] string extensionName, [NotNull
...
             var nativeSignatures = _profile.NativeSignatures
-                .Where(ns => ns.Extension == extensionName)
-                .Where(ns => ns.Categories.First() == category)
+                .Where(ns => ns.Extension == extensionName && ns.Categories.First() == category)
                 .OrderBy(s => s.Name)
                 .ThenBy(s => s.Parameters.Count)
                 .ToList();
 
             var overloads = _profile.Overloads
-                .Where(o => o.Extension == extensionName)
```
Comment (Nihlus) [inline src/Generator.Bind/Writers/BindingWriter.cs]: Why the combination here? It's more readable split up.

File: src/Generator.Bind/Writers/BindingWriter.cs [hunk 9]
```diff
@@ -193,7 +191,7 @@ private Task WriteExtensionWrapperFiles([NotNull] string extensionName, [NotNull
                 .ThenBy(s => s.ToString())
                 .ToList();
 
-            if (!categorySignatures.Any())
```
Comment (Nihlus) [inline src/Generator.Bind/Writers/BindingWriter.cs]: Same argument as the other place.

File: src/Generator.Bind/XML/Signatures/ApiProfile.cs [hunk 10]
```diff
@@ -11,6 +11,11 @@ namespace Bind.XML.Signatures
 {
     /// <summary>
     /// Represents an available API profile of a specific version.
+    ///
+    /// These are, broadly speaking, a collection of Function signatures, their overloads, and enums used by them.
```
Comment (Nihlus) [inline src/Generator.Bind/XML/Signatures/ApiProfile.cs]: Function with capital F?

File: src/Generator.Bind/XML/Signatures/Enumerations/EnumerationSignature.cs [hunk 11]
```diff
@@ -6,6 +6,8 @@ namespace Bind.XML.Signatures.Enumerations
 {
     /// <summary>
     /// Represents an enumeration definition in an API.
+    ///
+    /// This is a collection of values that functions can take. These are very much like enums in C#.
```
Comment (Nihlus) [inline src/Generator.Bind/XML/Signatures/Enumerations/EnumerationSignature.cs]: It's exactly an enum in C#. No need for the extra clarification.

File: src/Generator.Bind/XML/Signatures/Functions/FunctionSignature.cs (hunk 5 above)
Comment (varon) [reply src/Generator.Bind/XML/Signatures/Functions/FunctionSignature.cs]: `.Count > 0` is the correct usage of the object's API. given that we're not working working with an `IReadOnlyList<T>` and not just an `IEnumerable<T>`.

File: src/Generator.Bind/Program.cs (hunk 7 above)
Comment (varon) [reply src/Generator.Bind/Program.cs]: It's the correct GC mode for a batch tool. The default is designed for low-latency client-facing operations.

File: src/Generator.Bind/Writers/BindingWriter.cs (hunk 8 above)
Comment (varon) [reply src/Generator.Bind/Writers/BindingWriter.cs]: Because you're enumerating a collection twice for no reason.

Would you write?
```csharp
var someList = blahblah;
for (int i = someList.Count - 1; i >= 0; i --)
{
    if (predicate(someList[i]))
    {
        someList.RemoveAt[i];
    }
}
for (int i = someList.Count - 1; i >= 0; i --)
{
    if (anotherPredicate(someList[i]))
    {
        someList.RemoveAt[i];
    }
}
```

Or would you write:

```csharp
var someList = blahblah;
for (int i = someList.Count - 1; i >= 0; i --)
{
    if (predicate(someList[i]) || anotherPredicate(someList[i]))
    {
        someList.RemoveAt[i];
    }
}

Comment (jvbsl) [reply src/Generator.Bind/Writers/BindingWriter.cs]: Well LinQ does not iterate twice though with Where statements. as on each iterator the where statement yields if the predicate is true, but the next Where can skip it while it stll did not iterate through the whole Enumerable.
So it's not as bad as iterating twice, but performance wise it's ofc worse, because it allocates an additional Enumerator and delegate(not that bad for things like this) and has a double the function calls...
Readability wise - in my opinion it is debatable here as well which one is better.

File: src/Generator.Bind/XML/Signatures/Functions/FunctionSignature.cs (hunk 5 above)
Comment (DylanFPS) [reply src/Generator.Bind/XML/Signatures/Functions/FunctionSignature.cs]: I'm a LINQ-aholic, but I think varon's got this right. We should use `Count`.

File: src/Generator.Bind/Program.cs (hunk 7 above)
Comment (DylanFPS) [reply src/Generator.Bind/Program.cs]: Think of the numbers, mason.

(excuse the Black Ops 2 reference, but can we have some stats?)

File: src/Generator.Bind/Writers/BindingWriter.cs (hunk 8 above)
Comment (Nihlus) [reply src/Generator.Bind/Writers/BindingWriter.cs]: Yah, fair enough :)

File: src/Generator.Bind/XML/Signatures/Functions/FunctionSignature.cs (hunk 5 above)
Comment (Nihlus) [reply src/Generator.Bind/XML/Signatures/Functions/FunctionSignature.cs]: Alright :)

**Comment by Nihlus** [comment]
I'll admit, I'm a bit of a zealous LINQ fan, lol

Looks good overall. Need to see some stats on removing asynchronicity from the writing methods, and a better place for the license text, but other than that, I'm happy (provided the discussed changes are implemented).

File: src/Generator.Bind/Writers/BindingWriter.cs (hunk 4 above)
Comment (varon) [reply src/Generator.Bind/Writers/BindingWriter.cs]: These benchmarks are with all the changes. 

The file IO operation is serial, but we execute it for each file as it's own task in parallel.
I'm using a multi-core with all targets enabled and the performance has improved.

The reason why it's faster is very simple:
Writing each line via async/await is very inefficient. It would involve creating task objects, indirection and possibly context switches. It's a serial operation operation that should be done by one task (per file), not one task per line.

I don't have time to separate out the license changes and benchmark them for an offline tool, but it's pretty obvious that hitting the disk each time we write a file is not great for performance.

File: src/Generator.Bind/Program.cs (hunk 7 above)
Comment (varon) [reply src/Generator.Bind/Program.cs]: I swapped based on the fact that it's the correct mode for the type of processing we're doing.

But here are your numbers for interest sake:
Default time in GC: 764ms
Batch time in GC: 693ms

Default total time:  10.4340316
Batch total time: 10.3568002

Time in GC was profiled using dotTrace (timeline mode).
Total time was read from the app output with no profiling attached.

**Comment by varon** [comment]
This is all I have the time for atm, hopefully we can merge with this.

File: src/Generator.Bind/Program.cs (hunk 7 above)
Comment (Nihlus) [reply src/Generator.Bind/Program.cs]: Nice!

File: src/Generator.Bind/Writers/BindingWriter.cs (hunk 4 above)
Comment (Nihlus) [reply src/Generator.Bind/Writers/BindingWriter.cs]: I ran the tests myself, and it is indeed significantly faster. Sorry for being so difficult on this one :P Whenever there's a change that gets made for perf reasons, I'm always very interested to see why and how much of a difference there is.

**Comment by Nihlus** [comment]
There are still some issues with the documentation, and the issue with a settings object containing business data. We'll need to fix that before merging.

**Comment by varon** [comment]
Just merge and fix it later, it's definitely an improved state over the last version.
